### PR TITLE
[Pablo] Added fee tests for swap, buy, and remove liquidity [CU-390y2qt]

### PR DESCRIPTION
**Pull request by PoisonPhang**
[[CU-390y2qt]](https://app.clickup.com/t/390y2qt)

### CODE REVIEW

**Comment by ComposableFiB0t** [comment]
Task linked: [CU-390y2qt Test Non-Zero Fees](https://app.clickup.com/t/390y2qt)

File: code/parachain/frame/pablo/src/test/dual_asset_constant_product_tests_new.rs [hunk 1]
```diff
@@ -95,3 +95,258 @@ fn add_remove_lp() {
...
+
+			assert_ok!(Pablo::do_buy(&BOB, pool_id, BTC, usdt_to_buy, false));
+
+			// Fees deducted from buy amount in
+			assert!(default_acceptable_computation_error(
+				Tokens::balance(BTC, &BOB),
+				bob_btc - usdt_to_buy.amount - fee_amount
+			)
+			.is_ok());
+			assert_eq!(Tokens::balance(USDT, &BOB), usdt_to_buy.amount);
```
Comment (vimukthi-git) [inline code/parachain/frame/pablo/src/test/dual_asset_constant_product_tests_new.rs:151]: We need to assert also that the pool total liquidity has increased by the fee amount that was charged

File: code/parachain/frame/pablo/src/test/dual_asset_constant_product_tests_new.rs [hunk 2]
```diff
@@ -95,3 +95,258 @@ fn add_remove_lp() {
...
+
+	use super::*;
+
+	#[test]
+	fn should_deduct_fees_from_user_when_non_zero_fees() {
+		let initial_btc = 512_000_000_000;
+		let initial_usdt = 512_000_000_000;
+		let fee = Permill::from_rational::<u32>(3, 1000);
+		let bob_btc = 256_000;
+		let usdt_to_buy = AssetAmount::new(USDT, 128_000);
```
Comment (vimukthi-git) [inline code/parachain/frame/pablo/src/test/dual_asset_constant_product_tests_new.rs:111]: Let's define the variable close to where it's used

File: code/parachain/frame/pablo/src/test/dual_asset_constant_product_tests_new.rs [hunk 3]
```diff
@@ -95,3 +95,258 @@ fn add_remove_lp() {
...
+		// 50/50 BTC/USDT Pool with a 0.3% fee
+		let init_config = PoolInitConfiguration::DualAssetConstantProduct {
+			owner: ALICE,
+			assets_weights: dual_asset_pool_weights(BTC, Permill::from_percent(50), USDT),
+			fee,
+		};
+
+		new_test_ext().execute_with(|| {
+			process_and_progress_blocks::<Pablo, Test>(1);
+
```
Comment (vimukthi-git) [inline code/parachain/frame/pablo/src/test/dual_asset_constant_product_tests_new.rs:122]: Can we add a short comment about the test to explain?
Eg: Bob is buying <x> USDT with BTC with the fee paid in BTC

File: code/parachain/frame/pablo/src/test/dual_asset_constant_product_tests_new.rs [hunk 4]
```diff
@@ -95,3 +95,258 @@ fn add_remove_lp() {
...
+			// Mint tokens for buy
+			assert_ok!(Tokens::mint_into(BTC, &BOB, bob_btc));
+			assert_eq!(Tokens::balance(BTC, &BOB), bob_btc);
+
+			assert_ok!(Pablo::do_buy(&BOB, pool_id, BTC, usdt_to_buy, false));
+
+			// Fees deducted from buy amount in
+			assert!(default_acceptable_computation_error(
+				Tokens::balance(BTC, &BOB),
+				bob_btc - usdt_to_buy.amount - fee_amount
```
Comment (vimukthi-git) [inline code/parachain/frame/pablo/src/test/dual_asset_constant_product_tests_new.rs:148]: This doesn't look right why are we subtracting a usdt amount from a btc amount ?

File: code/parachain/frame/pablo/src/test/dual_asset_constant_product_tests_new.rs [hunk 5]
```diff
@@ -95,3 +95,258 @@ fn add_remove_lp() {
...
+	use super::*;
+
+	#[test]
+	fn should_deduct_fees_from_user_when_non_zero_fees() {
+		let initial_btc = 512_000_000_000;
+		let initial_usdt = 512_000_000_000;
+		let fee = Permill::from_rational::<u32>(3, 1000);
+		let bob_btc = 256_000;
+		let usdt_to_buy = AssetAmount::new(USDT, 128_000);
+		let fee_amount = fee.mul_ceil(usdt_to_buy.amount);
```
Comment (vimukthi-git) [inline code/parachain/frame/pablo/src/test/dual_asset_constant_product_tests_new.rs:112]: When buying USDT with BTC fee is in BTC, this calculation can't be correct

File: code/parachain/frame/pablo/src/test/dual_asset_constant_product_tests_new.rs [hunk 6]
```diff
@@ -95,3 +95,258 @@ fn add_remove_lp() {
...
+			min_receive()
+		));
+		assert_ok!(Pablo::remove_liquidity(
+			Origin::signed(ALICE),
+			pool_id,
+			alice_lpt_balance,
+			min_receive()
+		));
+
+		assert!(Tokens::balance(BTC, &ALICE) > initial_btc);
```
Comment (vimukthi-git) [inline code/parachain/frame/pablo/src/test/dual_asset_constant_product_tests_new.rs:334]: Can't we assert also the exact amounts?

File: code/parachain/frame/pablo/src/test/dual_asset_constant_product_tests_new.rs [hunk 7]
```diff
@@ -95,3 +95,258 @@ fn add_remove_lp() {
...
+	}
+
+	#[test]
+	fn should_distribute_fees_to_lps() {
+		let with_swaps = new_test_ext().execute_with(|| remove_liquidity_simulation(true));
+		let without_swaps = new_test_ext().execute_with(|| remove_liquidity_simulation(false));
+
+		// Ensure that fees are distributed to LPs by simulating remove liquidity with and without
+		// swaps
+		assert!(with_swaps.0 > without_swaps.0);
```
Comment (vimukthi-git) [inline code/parachain/frame/pablo/src/test/dual_asset_constant_product_tests_new.rs:349]: :+1:

File: code/parachain/frame/pablo/src/test/dual_asset_constant_product_tests_new.rs [hunk 8]
```diff
@@ -95,3 +95,258 @@ fn add_remove_lp() {
...
+		let alice_lpt_balance = Tokens::balance(lp_token, &ALICE);
+		let charlie_lpt_balance = Tokens::balance(lp_token, &CHARLIE);
+
+		// Mint tokens for swap
+		assert_ok!(Tokens::mint_into(BTC, &BOB, bob_btc));
+		assert_eq!(Tokens::balance(BTC, &BOB), bob_btc);
+		assert_ok!(Tokens::mint_into(USDT, &DAVE, dave_usdt));
+		assert_eq!(Tokens::balance(USDT, &DAVE), dave_usdt);
+
+		if with_swaps {
```
Comment (vimukthi-git) [inline code/parachain/frame/pablo/src/test/dual_asset_constant_product_tests_new.rs:302]: Need to consider buys as well

**Review by KaiserKarel — APPROVED** [review]

**Review by vimukthi-git — APPROVED** [review]
LGTM

**Pull request merged**